### DISCUSSION TITLE
[solved] Value by Reference issue with using an EntityChangeSubscriber: Finding one entity, changing one of its attribute values, and persisting and flushing it, updates all entities in that table.

### DISCUSSION THREAD

**Discussion by woltob** [mikro-orm/mikro-orm#3286] [Q&A]
**Describe the bug**
Running the following code updates all the entities and thus produces the following sql queries:

Am I doing something wrong? My understanding is it should only update the Entity with the id  = 7e78f41c-fbf1-420d-aa76-e4382c7e2464 that it found previously. 

```typescript 
    const organizationInvitation = await em.findOne(
      OrganizationInvitation,
      { id: id },
    );
    if (!organizationInvitation) {
      throw new NotFoundError('OrganizationInvitation');
    }
    if (!organizationInvitation.voidedAt) {
      organizationInvitation.voidedAt = new Date();
      await em.persistAndFlush(organizationInvitation);
    }
    return organizationInvitation;
```

```sql 
[query] select "o0".* from "organization_invitation" as "o0" where "o0"."deleted_at" is null and "o0"."id" = '7e78f41c-fbf1-420d-aa76-e4382c7e2464' limit 1 [took 17 ms]

[query] update "organization_invitation" set "voided_at" = '2022-07-06T15:02:25.818Z', "updated_at" = '2022-07-06T15:02:25.818Z' [took 2 ms]
```

*What is missing in the update where is the where `id = '7e78f41c-fbf1-420d-aa76-e4382c7e2464'`, therefore it should produce
```sql 

[query] update "organization_invitation" set "voided_at" = '2022-07-06T15:02:25.818Z', "updated_at" = '2022-07-06T15:02:25.818Z' where id = '7e78f41c-fbf1-420d-aa76-e4382c7e2464' [took 2 ms]
```

**Stack trace**
```
...
```

**To Reproduce**
Use the steps above. Perhaps follow the code implementation first. I can reproduce it. It started to happen after the latest updates, perhaps even my 5.2 update from 4.5.9.

**Expected behavior**
Changing an attribute on a single entity / object identified by id should only update this objects attributes. The update query for sql should have the id of the object as limiter.

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | 16.15.1 |
| typescript | 4.3.5 |
| mikro-orm | 5.2.1 |
| your-driver | @mikro-orm/postgresql |

**Comment by woltob** [comment]
Note: Following the official documentation with wrap(myEntity).assign({attr: "new value"}); and flushing it produces also an update query without any limitation.

```typescript 

const organizationInvitation = await this.getEm().findOne(
      OrganizationInvitation,
      { id: id },
    );
    if (!organizationInvitation) {
      throw new NotFoundError(
        `OrganizationInvitation not found`,
        `OrganizationInvitation with id ${id} was not found.`,
      );
    }
    if (!organizationInvitation.voidedAt) {
      wrap(organizationInvitation).assign({ voidedAt: new Date() });
      await this.getEm().flush();
    }
    return organizationInvitation;
```

Produces:

NOTE: the deleted_at column is for soft delete and is a default filter that is always on via a custom BaseEntity. These update queries used to work before 5.2 with GraphQL and NestJS.
```sql
[query] select "o0".* from "organization_invitation" as "o0" where "o0"."deleted_at" is null and "o0"."id" = '7e78f41c-fbf1-420d-aa76-e4382c7e2464' limit 1 [took 2 ms]


[query] update "organization_invitation" set "voided_at" = '2022-07-06T15:26:47.693Z', "updated_at" = '2022-07-06T15:26:47.696Z' [took 3 ms]
```

**Comment by woltob** [comment]
Just a hint and perhaps a debugging starting point.

Note that I am using a BaseEntity where I auto assign a v4 uuid to the id attribute.
In order to have the em.create({}) not complain about auto-assigned attributes with default values, I added a
```
[OptionalProps]?: 'id'
```
into the BaseEntity from which all Entities inherit.

Could this cause the generated update function to ignore the id field?

**Comment by woltob** [comment]
Another observation: 

Using the query builder, it successfully uses the id as a where limiter as defined. However, the default value onUpdate => new Date() which should be set for the updatedAt column in my BaseEntity is ignored. Secondly, the result variable from execute is an empty array (is that correct or would the entity/entities be returned?). 

```typescript 

const qb = em.createQueryBuilder(OrganizationInvitation);
      const result = await qb
        .update({ voidedAt: new Date() })
        .where({ id })
        .execute();
      console.log('!!!!! result', result);
```

```
[query] select "o0".* from "organization_invitation" as "o0" where "o0"."deleted_at" is null and "o0"."id" = '7e78f41c-fbf1-420d-aa76-e4382c7e2464' limit 1 [took 0 ms]
[query] update "organization_invitation" set "voided_at" = '2022-07-06T15:45:13.503Z' where "id" = '7e78f41c-fbf1-420d-aa76-e4382c7e2464' [took 1 ms]
!!!!! result []
``` 

My current workaround is:

```typescript

if (!organizationInvitation.voidedAt) {
      const qb = em.createQueryBuilder(OrganizationInvitation);
      const result = await qb
        .update({ voidedAt: new Date(), updatedAt: new Date() })
        .where({ id })
        .execute();
      console.log('!!!!! result', result);
      return (await em.findOne(OrganizationInvitation, {
        id,
      })) as OrganizationInvitation; // because I cannot return null.
}
return organizationInvitation;
```

Any suggestions how to solve this problem? I would have to change my entire code base before being able to go live with version 5.2, as I am updating other entities everywhere else as described in my first bug descriptions.

Thank you very much for a quick answer.

**Comment by woltob** [comment]
After updating to 5.2.2, the problem still persists. This has not happened on version 4.5.x. Thankful for any hints.

**Comment by woltob** [comment]
The problems persist also with using EntityRepository. Update queries are generated without a limiting where condition. Please have a look, I hope I am wrong. It was not happening a week ago before migrating to v5.2 from 4.5.10.

**Comment by woltob** [comment]
Another thing I am noticing:

I need to explicitly assign `...jobOne.mainJob` and `...jobOne.otherJob` to the relational attributes mainJob: and otherJob respectively so that existing nested attributes are not omitted. It does not work without e.g. `...jobOne.mainJob` before `...mainJobInput` input.  All these queries worked before with v4.5.10. Am I doing something wrong?

```typescript

wrap(jobOne).assign(
        {
          ...jobOneInput,
          mainJob: {
            ...jobOne.mainJob,
            ...mainJobInput,
          },
          otherJob: {
            ...jobOne.otherJob,
            ...otherJobInput,
          },
        },
        { mergeObjects: true }
```

Postgres, latest Mikro ORM v5.2.2, NestJS, PostgreSQL, Typescript.

**Comment by B4nan** [comment]
I am sorry but this is a mess, please provide a complete but minimal reproduction. 

https://stackoverflow.com/help/minimal-reproducible-example

**Comment by woltob** [comment]
Is there a repository with a minimal reproduction example that I can adjust?

My problem is, that since the upgrade to version 5.2, the process of simply updating an attribute would generate a sql query (postgres) that does not have a limiting where condition. Hence, it updates all the entries in the database, as described in my first post under this discussion.

**Reply by woltob** [reply]
Like I mentioned, all entities inherit from a BaseEntity (custom) that holds id, createdAt, updatedAt, etc. and the attributes which have default values are ignored by setting them with `[OptionalProps]?:` as per your documentation. Could this have an impact on the generation of the update queries? I am pretty sure that you have tested this standard case of generating valid update sql queries with correct where conditions. Thank you.

```typescript

export type OPTIONAL_BASE_ENTITY_ATTRIBUTES_TYPE =
  | 'id'
  | 'createdAt'
  | 'updatedAt'
  | 'deletedAt';
@ObjectType({ isAbstract: true }) // GraphQL
@InputType({ isAbstract: true }) // GraphQL
@Entity({ abstract: true }) // MikroORM
@Directive('@extends') // GraphQL
@Directive('@key(fields: "id")') // GraphQL

export abstract class BaseEntity<O = OPTIONAL_BASE_ENTITY_ATTRIBUTES_TYPE> {
  // O is set to be OPTIONAL_BASE_ENTITY_ATTRIBUTES_TYPE by default.
  // When O is set however, we make sure that O & OPTIONAL_BASE_ENTITY_ATTRIBUTES_TYPE are optional.
  [OptionalProps]?: OPTIONAL_BASE_ENTITY_ATTRIBUTES_TYPE | O;

  @Field((_type) => ID, { nullable: false, defaultValue: null }) // GraphQL
  @PrimaryKey({
    type: 'uuid',
    nullable: false,
    defaultRaw: 'uuid_generate_v4()',
  }) // NestJs / MikroORM
  public id: string = v4();

.... 
```

**Reply by woltob** [reply]
PS: I just highlighted the code snippets with the right syntax (typescript, sql) for readability. Please have a quick look at it as you would probably know what could be the cause of this.

**Reply by B4nan** [reply]
Check this folder for inspiration, ideal reproduction would be a single file like those. 

https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

Note that you dont have to make this into a test, you dont have to clone this repo and add a test in there, it can be a regular (standalone) script. But the contents should be quite the same. Ideally no graph ql, I can't really help with that. Reproductions should be minimal, so ideally no other dependencies than the ORM.

> Could this have an impact on the generation of the update queries? 

Nope, that code seems totally fine to me. As I mentioned above, I kinda doubt the Unit of Work would produce such query. Updates are issued only for managed entity, and we validate the entity has PK before it can be managed.

**Comment by woltob** [comment]
I found the issue.

I have a subscriber (https://mikro-orm.io/docs/next/events) with the Mikro ORM configuration that logs every change to entities to the database. Somehow this subscriber was altering the update queries. 

By removing this line from the Mikro ORM configuration the queries were working properly again. 

`subscribers: [new EntityChangeSubscriber()]`

What I need: A proper way of logging insert and update events to any entity.

**Reply by B4nan** [reply]
How does the `EntityChangeSubscriber` definition look like?

**Reply by woltob** [reply]
It was not copying the changeSets but rather using them by reference.

```typescript

async afterFlush(args: FlushEventArgs): Promise<void> {
    console.log('===========7777 afterFlush', this.historyLogTransactionId);
    // console.log('args', args);
    const changeSets = [...args.uow.getChangeSets()];  // NOTE: here I create a copy.
    const entities = changeSets.map((cs) => cs.entity);
    console.log('changeSet', changeSets);
    console.log('entities', entities);

    for (let cs of Object.values(changeSets)) {
      if (
        cs.name !== 'HistoryLog' &&
        // cs.name !== 'DeviceLog' &&
        cs.name !== 'AvailabilityLocation' &&
        cs.name !== 'DeviceLocation'
      ) {
        console.log('print changeset');
        const nextValues = cs.payload ? { ...cs.payload } : {};
        const tableName = cs.name;
        const operation =
          HistoryLogChangeSetType[cs.type.toString().toUpperCase()] ===
          undefined
            ? HistoryLogChangeSetType.UNKNOWN_ERROR
            : HistoryLogChangeSetType[cs.type.toString().toUpperCase()];

        const rowId: string = cs.getPrimaryKey() as string;
        const modifiedFields = [...Object.keys(cs.payload)];
        const entityValues = cs.originalEntity ? { ...cs.originalEntity } : {};
        const previousValues = { ...entityValues };
        // Iterating over all ()
        for (let key in Object.keys({ ...entityValues })) {
          if (
            !entityValues.hasOwnProperty(key) ||
            !modifiedFields.includes(key)
          ) {
            delete previousValues[key];  // NOTE: perhaps this was causing the issue in v5.x
          }
        }

        console.log('previousValues', entityValues);
        console.log('modifiedFields', modifiedFields);
        console.log('nextValues', nextValues);
        console.log('tableName', tableName);
        console.log('operation', operation);

        const forkedEm: SqlEntityManager<PostgreSqlDriver> = args.em.fork({
          clear: true,
        }) as SqlEntityManager<PostgreSqlDriver>;

        const historyLog = forkedEm.create(HistoryLog, {
          id: v4(),
          operation,
          tableName,
          rowId,
          previousValues,
          nextValues,
        });

        const qb = forkedEm.createQueryBuilder(HistoryLog);
        try {
          const result = await qb
            .insert(historyLog)
            .onConflict('id')
            .ignore()
            .execute('get');
          console.log('historyLog result', result);
        } catch (e) {
          console.error('Catching Error from insert HistoryLog... ignoring');
          console.error(e);
        }
      }
    }
  }

```

**Comment by woltob** [comment]
Another thing (unrelated) I just noticed. The generation of keys for Postgresql is wrong. The long names get chopped by postgres but migrations try to generate them again. Particularly:

```
  jse_cause: TableExistsException: alter table "organization_organization" add constraint "organization_organization_used_organization_invitation_id_unique" unique ("used_organization_invitation_id"); - relation "organization_organization_used_organization_invitation_id_uniqu" already exists
````

```sql

    this.addSql(
      'alter table "organization_organization" add constraint "organization_organization_used_organization_invitation_id_unique" unique ("used_organization_invitation_id");',
    );
```

**Reply by woltob** [reply]
Not that `organization_organization_used_organization_invitation_id_unique` becomes `organization_organization_used_organization_invitation_id_uniqu` (e in the end is missing).

**Reply by B4nan** [reply]
Please provide separate reproduction for this. There are tests and they pass:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/schema-generator/max-index-name-length.postgres.test.ts
https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/schema-generator/__snapshots__/max-index-name-length.postgres.test.ts.snap

(entity definition should be enough here)

**Reply by woltob** [reply]
Note: it is note a foreign key (I have seen that shortening happening there) but a unique key which runs also as a constraint in postgresql.

**Reply by B4nan** [reply]
The test I linked above also defines a unique constraint, not just FK.

**Comment by woltob** [comment]
Another thing:

When running `npx mikro-orm migration:create` I have to delete the `./dist` (compiled NestJS files) first in order for the migration generation to reflect the current state of the typescript source files. Is there a way to ignore that and always check against the source files or at least run a new build process beforehand (without having to generate a package.json script for that)? I guess other people would find this as well and wonder why changes are not reflected. Thanks!

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/migrations#snapshots

**Reply by woltob** [reply]
thanks!

**Comment by woltob** [comment]
Another question:

I put the migration runner into my AppModule as per 

```typescript 
@Module({
....
})
export class AppModule implements NestModule, OnModuleInit {
  constructor(private readonly orm: MikroORM) {}

  // New way of running migrations. Enable if the rest works and disable them from the cluster perhaps.
  async onModuleInit(): Promise<void> {
    console.log(
      'app.module.ts :: AppModule :: onModuleInit() :: Running MikroORM migrations here.',
    );
    await this.orm.getMigrator().up();
  }

  // for some reason the auth middlewares in profile and article modules are fired before the request context one,
  // so they would fail to access contextual EM. by registering the middleware directly in AppModule, we can get
  // around this issue
  configure(consumer: MiddlewareConsumer) {
    consumer.apply(MikroOrmMiddleware).forRoutes('*');
  }
}
```

however I am receiving 

```typescript
import { Migration } from '@mikro-orm/migrations';
  ^^^^^^
  
  SyntaxError: Cannot use import statement outside a module
```

What is the most common error here?

**Reply by woltob** [reply]
When there are pending migrations and they run when starting my NestJS application.

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/migrations#running-migrations-in-production
